### DISCUSSION TITLE
UI gets suspended in mutations since upgrade to v2

### DISCUSSION THREAD

**Discussion by carlesba** [pmndrs/jotai#1841] [Q&A]
# context
- The app uses react-query (with jotai atoms) to cache Server state.
- The app holds those queries under `jotai-molecules` (may not be relevant)
- Due to this structure, there're many derived atoms that are async (as source is async). This provoked a large amount of changes in all derived atoms creating many `atom(get => get(otherAtom).then(s => s.value))`

# case
- When react-query's query-client gets updated in atom mutation, all components reading the updated `query` trigger suspense.
This didn't happen before with `jotai v1`. 
- **This change is synchronous**. However, since is changing the async atoms state, there're many atoms updated as side effect.
- The suspense can be prevented by wrapping the cache update in a react's `startTransition`. However, the update is still very slow (way more than jotai v1) and the following warning is prompt in console: 

```
Warning: Detected a large number of updates inside startTransition. If this is due to a subscription please re-write it to use React provided hooks. Otherwise concurrent mode guarantees are off the table.
```

- None of this happens when react-query's cache is invalidated. This means, instead of setting `queryClient` data manually, tell react-query to refetch data and update then queryClient.

# Question
Is there any flaw in jotai v2 that make atom subscription slower with async atoms? 
Is there any recommendation to generate derived atoms in a more performant way?
Is this an issue with react-query/jotai integration?

**Comment by dai-shi** [comment]
> Is there any flaw in jotai v2 that make atom subscription slower with async atoms?

Interesting. Don't know...
As you see v1 and v2 behave very differently for async atoms. Would you able to try creating a very small reproduction hopefully without query integration?

**Reply by carlesba** [reply]
Here, this describes the use case:

https://codesandbox.io/s/jotai-react-query-mutation-trigger-suspense-on-set-queryclient-0b7fou


# case A
1. uncheck "start transition" 
2. click on update
The values will be unmounted despite being a synchronous operation

# caseB
1. check "start transition"
2. click on update
The values don't unmount but there's a warning in log suggesting there are many subscriptions

---

What do you think? Is this expected? What should be the best practice?

**Reply by carlesba** [reply]
I think the guilty is here
https://github.com/jotaijs/jotai-tanstack-query/blob/main/src/common.ts#L129

This could make the operation async and trigger the suspense. However, I'm concerned about the warning. Do you think is due to `useAtom` subscriptions?

**Reply by dai-shi** [reply]
Yeah, that `Promise.resolve()` is a hack, because RQ notifies us in sync. I don't know how to solve it otherwise. I don't think it's jotai issue, unless we have a repro without jotai-tanstack-query.

**Reply by carlesba** [reply]
Let me get this. When RQ notifies using `subscribe` callback, what's the use of `Promise.resolve`? Cannot we just call the `next` provided in the callback to get the result?

I'm aware this may not be the issue, but I'd like to understand, maybe I created another jotai wrapper for RQ.
Thanks for clarifying.

**Reply by dai-shi** [reply]
If you remove it (`Promise.resolve()`), you will see React warning. (I thought you already saw it.)

**Comment by dai-shi** [comment]
If my assumption is correct, there's a workaround in `useAtom`.

You can specify `delay` option. I think in your case, it can be `0`.

https://github.com/pmndrs/jotai/blob/ab1e7b641447d63bb23efa2925bbbf954b28886a/src/react/useAtomValue.ts#L45

**Reply by carlesba** [reply]
this seems to work on the codesandbox but not on my codebase :/ 
Thanks for the suggestion, anyway.

What's the use case for `delay`? Does it work as a `deferred value`?

**Reply by dai-shi** [reply]
It should be a different thing. It's just `setTimeout`.
https://github.com/pmndrs/jotai/blob/ab1e7b641447d63bb23efa2925bbbf954b28886a/src/react/useAtomValue.ts#L92

**Comment by carlesba** [comment]
I may figured out the reason for this suspense triggering in my codebase. Not sure why this can't be detected in codesandbox's playground.

The codebase I'm working with has 2 or 3 async atoms (query atoms) that hold the server state. There are plenty of derived atoms that are based on those async queries and also on other derived atoms. 

Since V2 has introduced that `get` need to be `await`ed, many of the derived atoms have become a composition of multiple levels of await. This seems to be the reason that triggers suspense when the source changes (even synchronously). The source changes and then man atoms will have to `await` `await` `await` till data is reached.

Maybe I'm missing some point and I'm composing atoms in the wrong way.

I see the benefits that jotai v2 brought but after this issue, I feel now jotai way more slower. Every derived atom (that before was even beneficial for atomic operations and lazy evaluation) has now the drawback of delaying the computation.

**Reply by dai-shi** [reply]
yeah, v2 is completely different design for async atoms.
fwiw, v1 isn't very fast either, because all atom got invalidated and then waited to re-compute.
what's different is, in v1, it bails out behind the scene.
in v2, sync atoms are actually faster than v1, but async atoms naively can be slower.

So, the idea is we need to optimize it by ourselves. `unwrap` is one solution, but doesn't work if you want suspense.

You are reaching a great point. I wonder if you can provide a small example to play with.
Something like `atomWithCache` approach might help.

**Reply by carlesba** [reply]
Even `atomWithCache` example is throwing a little suspense for the cached atom. Even being small, has dramatically side effects in bigger apps.

I'm not sure about the solution. How I see the problem is:

```
const server = // async atom
const local = // sync atom

const derivedData = // use async first time
// local later to avoid sync (local needs to be updated when server is resolved)
```

Problem is that derived data will have to be async just because of initial state is async, then all the derived atoms based on `derived` are adding more async. 

I'm missing some sort of reactivity that allows subscription from async atoms but only trigger suspense on first load and updates state without suspense later. Not sure if it's feasible but I was getting that before with `jotai v1` and I also with plain `react-query`. 

Do you see a quick work around for these?

**Reply by dai-shi** [reply]
Not sure if I get your problem 100% correctly, but we shouldn't use async, but promise instead:
```js
const asyncAtom = atom(async (get) => ...)
const syncAtom = atom(null)

// this will be async, but then sync if syncAtom is set.
const derivedAtom = atom(
  (get) => get(syncAtom) || get(asyncAtom)
)
```

**Reply by carlesba** [reply]
yet, we'll have to drag that `value instanceof Promise ? value.then(readData) : readData(value)` to every derived atom, right?

**Reply by carlesba** [reply]
that's especially challenging when `readData` depends on 2 async atoms

```js
const remoteAAtom;// async 
const remoteBAtom; // async

const resultAtom = atom(get=> 
  Promise.all([
   get(remoteAAtom), 
   get(remoteBAtom)
]).then(([a , b] ) => readData(a , b) )
```

**Reply by dai-shi** [reply]
I'm pretty sure you get the point and we are on the same page.
selectAtom may help, but that also means you'll have to rewrite all async atoms.

Another workaround is to use a few base async atoms higher in the component tree and once they are resolved, we show children. There all derived async atoms can be in sync (with `unwrap`).

**Comment by ratiofu** [comment]
With the latest versions of React (18.2.0) & Jotai (2.6.0) at the time of this writing I also see the warning

```
Warning: Detected a large number of updates inside startTransition.
```

in [the public "tearing" example](https://codesandbox.io/p/sandbox/wqwwqt). Reducing the number of listeners/Counters to 10 or fewer makes the warning disappear. I presume the warning happens with more than 10 subscribers to a Jotai atom. Pulling the reading of the counter state up into the parent component of course also solves it and also makes the tearing disappear. I guess this is all working as designed, then, right?

**Reply by dai-shi** [reply]
I think so. But, if the warning shows in a real app, there might be some slow components and they should fix some app code.
Is it correct, if you remove the while loop, it won't show the warning?
